RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+  before the final allocation of the output tensor. You could save a few
+  instructions by eliminating this struct entirely. However, I could not
+  think of an elegant API for doing so.
+
+In absolute terms, the boilerplate code saved here is not all that
+large; only several lines. However, when multiplied over the number of
+kernels in PyTorch, and the subtlety of remembering to handle all of
+these issues, and we think the use of code generation to automatically
+generate this boilerplate is worth it.
+
+The functionality code generated here is not made available to other
+backends. This is problematic, so we also generate a separate dispatch
+key handler which will automatically handle shape checking and other
+concerns for backends that are not concerned with performance:
+
+```
+// Name TBD; for this example we will call it DispatchKey::Common, as the
+// functionality here is common to all backends. This is an alias key
+// that resolves CommonXLA/CommonMSNPU/... in the same way as Autograd.
+
+Tensor upsample_nearest1d_common(const Tensor& self, IntArrayRef output_size, optional<double> scales) {
+  // TODO: RecordFunction could be added here, if desired
+  DeviceGuard g(self.device());
+  TensorMeta expected = meta::upsample_nearest1d(self, output_size, scales);
+  Tensor result = at::empty(expected.sizes(), expected.options());
+  ExcludeDispatchKeyGuard g2(DispatchKey::Common);
+  // Notice that upsample_nearest1d is ignored here. It may be a good idea
+  // to skip this implementation, or use a slightly different variant, if
+  // a backend explicitly registered upsample_nearest1d
+  at::upsample_nearest1d_out(expected, result, self, output_size, scales);
+  return result;
+}
+
+TORCH_LIBRARY_IMPL(aten, Common, m) {
+  m.impl("upsample_nearest1d", upsample_nearest1d_common);
+}
+```
+
+A backend that would like to fuse this boilerplate into their kernel for
+performance reasons can simply override the Common entry with
+fallthrough:
+
+```
+TORCH_LIBRARY_IMPL(aten, CommonXLA, m) {
+  m.impl("upsample_nearest1d", CppFunction::makeFallthrough);
+}
+```
+
+Finally, structured kernels also implicitly register an implementation
+for the Meta key, which is the API for dry running shape/dtype
+calculations without any kernels involved:
+
+```
+Tensor upsample_nearest1d_meta(const Tensor& self, IntArrayRef output_size, optional<double> scales) {
+  TensorMeta expected = meta::upsample_nearest1d(self, output_size, scales);
+  return native::empty_meta(expected.sizes(), expected.options());
+}
+
+TORCH_LIBRARY_IMPL(aten, Meta, m) {
+  m.impl("upsample_nearest1d", upsample_nearest1d_meta);
+}
+```
+
+## Discussion
+
+* Why not just use the Common dispatch key for everything?
+
+    * Performance reasons. Introducing the common key would induce an
+      extra redispatch, which at time of writing would give up quite a
+      bit of performance due to dispatch overhead, for no particularly
+      good reason.
```

Comment:
I think it's worth thinking carefully about whether this is the best way to make structured kernels available out of tree, given the modality it introduces, risk of drift, perf handicap (albeit diminishing over time) it saddles out of tree backends with, etc.

E.g. once the rest of the codegen has been ported, it's not absurd to imagine fitting it with a frontend that can take inputs other than the in-tree yaml...

Reply:
The common dispatch key is primarily oriented towards backend implementers.

Let's suppose for a moment, that an overriding design goal is safety/correctness first, with the ability to opt into performance.  Then it feels feel like I am forced to introduce the Common dispatch key, because without it, an average backend overrider has to faithfully replicate all of the functionality that we have otherwise fused into CPU/CUDA operators (shape checking, device guards, and with some refactors coming soon, it will also include version counter bumps). It's unrealistic to expect a backend implementer to actually manage all of this without code generation.

That being said, there is a certain optionality to the Common dispatch key. We don't have to implement it (and indeed, in the current posted PR, it is not implemented), and if it is not implemented, the burden is simply on backend implementers to implement all of the necessary scaffolding (which is the de facto situation today). If, for example, we pivoted to publishing code generation for backend implementers, that would alleviate or perhaps eliminate the need for a common dispatch key.

The common dispatch key is mostly irrelevant for custom operators, since average use of common operators is via catch all registration and there will not be any common dispatch key in any case.

Reply:
Yeah, agree on all the motivations. My pitch would be to (wait and) audition a lightly parameterized codegen pipeline as an alternative, once the rewrite is complete.